src/test/mod.rs

```diff
@@ -20,50 +20,67 @@ use env_logger;
 
 use analysis;
 use build;
-use server as ls_server;
+use server::{self as ls_server, ServerMessage, Request, Method};
 use vfs;
 
 use self::types::src;
 
 use url::Url;
-use ls_types::TextDocumentIdentifier;
+use ls_types::*;
 use serde_json;
 use std::path::{Path, PathBuf};
 
 const TEST_TIMEOUT_IN_SEC: u64 = 10;
 
+impl ServerMessage {
+    pub fn simple_test_initialize(id: usize, root_path: Option<String>) -> ServerMessage {
+        ServerMessage::Request(Request {
+            id: id,
+            method: Method::Initialize(InitializeParams {
+                process_id: None,
+                root_path: root_path,
+                root_uri: None,
+                initialization_options: None,
+                capabilities: ClientCapabilities {
+                    workspace: None,
+                    text_document: None,
+                    experimental: None,
+                },
+                trace: TraceOption::Off,
+            })
+        })
+    }
+}
+
 #[test]
 fn test_goto_def() {
     let (mut cache, _tc) = init_env("goto_def");
 
     let source_file_path = Path::new("src").join("main.rs");
 
-    let root_path = format!("{}", serde_json::to_string(&cache.abs_path(Path::new(".")))
-                                      .expect("couldn't convert path to JSON"));
+    let root_path = cache.abs_path(Path::new("."));
     let url = Url::from_file_path(cache.abs_path(&source_file_path)).expect("couldn't convert file path to URL");
-    let text_doc = serde_json::to_string(&TextDocumentIdentifier::new(url)).expect("couldn't convert path to JSON");
-    let messages = vec![Message::new("initialize", vec![("processId", "0".to_owned()),
-                                                        ("capabilities", "{ \"experimental\": null }".to_owned()),
-                                                        ("rootPath", root_path),
-                                                        ("rootUri", "null".to_owned()),
-                                                        ("trace", "\"off\"".to_owned())]),
-                        Message::new("textDocument/definition",
-                                     vec![
-                                        ("textDocument", text_doc),
-                                        ("position", serde_json::to_string(&cache.mk_ls_position(src(&source_file_path, 22, "world"))).expect("couldn't convert path to JSON"))
-                                     ])];
+
+    let messages = vec![
+        ServerMessage::simple_test_initialize(0,root_path.as_os_str().to_str().map(|x| x.to_owned())),
+        ServerMessage::Request(Request { id: 11, method: Method::GotoDefinition(TextDocumentPositionParams {
```

Comment:
Would something like:
```
fn request(id: usize, method: Method) -> ServerMessage {
    ServerMessage::Request(Request { id: id, method: method })
}
```
e.g.
```
ServerMessage::request(11, Method::GotoDefinition(TextDocumentPositionParams {
    text_document: text_doc,
    position: pos
}))
```
be good enough (also asking about style)?
I think it strikes a good balance between reducing redundancy and simplification while allowing to express every request type.
To get rid of explicit need to pass method params struct type I'd have to enumerate over variants (which I don't know how to do yet - I assume it needs macros 1.1 like [here](https://cbreeden.github.io/Macros11/)?) to get their payload types.